File: src/code/java/_078.java
package code.java;

import java.util.ArrayList;
import java.util.List;

public class _078 {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> lists = new ArrayList<>();
        lists.add(new ArrayList<Integer>());
        for (int num : nums) {
            for (int i = 0, size = lists.size(); i < size; i++) {
                List<Integer> list = new ArrayList<>(lists.get(i));
                list.add(num);
                lists.add(list);
            }
        }
        return lists;
    }
}

File: src/code/js/_078.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
    var lists = [[]];
    for (var i = 0; i < nums.length; i++) {
        for (var j = 0, size = lists.length; j < size; j++) {
            var list = lists[j].slice();
            list.push(nums[i]);
            lists.push(list);
        }
    }
    return lists;
};
/*
initial [ [] ]
add 1   [ [], [1] ]
add 2   [ [], [1], [2], [1,2] ]
add 3   [ [], [1], [2], [1,2], [3], [1,3], [2,3], [1,2,3] ]
 */